### ISSUE TITLE
Tile expiry problems - How can I override the expiry handling ?

### ISSUE THREAD

**Issue by pasniak** [osmdroid/osmdroid#1485]
Issue Type

[x] Question
[ ] Bug
[ ] Improvement

**TLDR**: My new tile overlay does not show fresh images. How can I override the expiry handling to use HTTP_EXPIRES_HEADER? BTW: Wouldn't it be nice to log the HTTP headers if isDebugMapTileDownloader = true? 

```
            final long expirationTime = computeExpirationTime(
                    c.getHeaderField(OpenStreetMapTileProviderConstants.HTTP_EXPIRES_HEADER),
                    c.getHeaderField(OpenStreetMapTileProviderConstants.HTTP_CACHECONTROL_HEADER),
                    System.currentTimeMillis());
```
Longer background:

In my new online tile source overlay (which url has time changing every hour...) I saw from the server a kind of non-standard response **max-age=[big-number], ... max-age=[small number]**. 

I debugged the code and I see we pick the first **[big-number]** but the server returns _both_ max-age and Expire time; yet we pick the **max-age**... I tried to override the `TileDownloader` as follows but I get no callbacks in `MyTileDownloader`. What am I doing wrong?

```
                mapView?.run {
                    v("adding tile downloader")
                    cacheManager = CacheManager(this).apply {
                        setTileDownloader(MyTileDownloader())
                    }
```

See also #1173 

6.1.6 (fairly recent master)

**Comment by pasniak** [comment]
Given there can be multiple TileSources from different servers, shouldn't downloadTile get its configuration (ex: getExpirationOverrideDuration) from TileSourcePolicy ? And DefaultConfigurationProvider.save/load store preferences per tileSource?

I added some HTTP debugging. Here's what I get from the servers as of  15 Jan 5:18 GMT:
**server 1**: success, mime is image/png, expires=Thu, 16 Jan 2020 10:18:53 GMT, cacheControl=public, max-age=259200, s-maxage=259200, max-age=86400
**server 2**: success, mime is image/png, expires=Mon, 20 Jan 2020 10:18:53 GMT, cacheControl=max-age=432000

**Comment by monsieurtanuki** [comment]
@pasniak This is what I would do:
* build my own `MapTileProviderBasic`
* override `TileDownloader.computeExpirationTime(...)` in order to use HTTP_EXPIRES_HEADER first (for instance)
* use `MapTileDownloader.setTileDownloader(TileDownloader)`

**Comment by pasniak** [comment]
Thanks! That works.  Would it be possible to pass pTileSource to computeExpirationTime (as additional arg) in TileDownloader.downloadTile so I could distinguish between tile sources? 

I'd rather overall subclass as I do not want to pull in too much downloading logic so I might have a small PR...

**Comment by monsieurtanuki** [comment]
@pasniak I'm not sure I understand why you would need the extra pTileSource parameter in `TileDownloader.computeExpirationTime`, as each instance of `MapTileDownloader` has a unique tile source and a unique `TileDownloader`: nothing prevents you now from using each tile download provider with its own expiration logic.

That being said, I guess that would make sense to use instead a new `OnlineTileSourceBase.computeExpirationTime` method (because it's the expiration algorithm is inherent to the tile source), to be used in `TileDownload.downloadTile` (which already has an `OnlineTileSourceBase` parameter).
Or perhaps put that in `TileSourcePolicy`.

What do you think of that?

**Comment by monsieurtanuki** [comment]
@pasniak I've just pushed #1533. Would that suit your needs?

**Comment by pasniak** [comment]
TileSourcePolicy works great - although it requires overriding entirety of computeExpirationTime just to set extension...

**Comment by spyhunter99** [comment]
isn't there some expiration settings on the config provider?
